---
source: Rmd
title: "Next steps"
teaching: XX
exercises: XX
questions:
- "SummarizedExperiment"
objectives:
- Introduce the notion of data containers
- Give an overview of the `SummarizedExperiment`, extensively used in omics analyses
keypoints:
- "`SummarizedExperiment` represent an efficient way to store and to handle omics data"
- "SummarizedExperiment is used in many Bioconductor packages"
- "We can use tidySummarizedExperiment to interact with SummarizedExperiment using tidyverse commands"
---

```{r, include=FALSE}
source("../bin/chunk-options.R")
```

## Next steps

Data in bioinformatics is often complex.
To deal with this, developers define specialised
data containers (termed classes) that match the properties of the
data they need to handle.

This aspect is central to the **Bioconductor**[^Bioconductor] project which
uses the same **core data infrastructure** across packages. This certainly contributed
to Bioconductor's success. Bioconductor package developers are advised to make
use of existing infrastructure to provide coherence, interoperability and stability to
the project as a whole. You will learn more about Bioconductor in the next episode.


[^Bioconductor]:The [Bioconductor](http://www.bioconductor.org) was initiated by Robert Gentleman, one of the two creators of the R language. Bioconductor provides  tools dedicated to omics data analysis. Bioconductor uses the R statistical programming language, and is open source and open development.


To illustrate such an omics data container, we'll present the `SummarizedExperiment`
class.

## SummarizedExperiment

The figure below represents the anatomy of SummarizedExperiment.

```{r SE, echo=FALSE, out.width = '80%'}
knitr::include_graphics("https://uclouvain-cbio.github.io/WSBIM1322/figs/SE.svg")
```


Objects of the class SummarizedExperiment contain :

- **One (or more) assay(s)** containing the quantitative omics data
  (expression data), stored as a matrix-like object. Features (genes,
  transcripts, proteins, ...) are defined along the rows and samples
  along the columns.

- A **sample metadata** slot containing sample co-variates, stored as a
  data frame. Rows from this table represent samples (rows match exactly the
  columns of the expression data).

- A **feature metadata** slot containing feature co-variates, stored as
  data frame. The rows of this dataframe's match exactly the rows of the
  expression data.
  

### Benefits of SummarizedExperiment format


#### Coordination of samples and features

The coordinated nature of the SummarizedExperiment guarantees that
during data manipulation, the dimensions of the different slots will always match (i.e the columns
in the expression data and the rows in the sample metadata, as well
as the rows in the expression data and feature metadata) during data
manipulation. For example, if we had to exclude one sample from the assay,
it would be automatically removed from the sample metadata in the same operation.

The metadata slots can grow additional co-variates
(columns) without affecting the other structures.

#### Interoperability between packages

There are a lot of R packages that make use of SummarizedExperiment format. If you follow a training focused on RNA sequencing analysis, you may learn to use the Bioconductor `DESeq2` package to do some differential expression analyses. `DESeq2`'s whole analysis is handled in a `SummarizedExperiment`. Or if you perform RNA sequencing analysis following tidy principles with the [tidybulk](https://stemangiola.github.io/tidybulk/) package you can input your data in SummarizedExperiment format. 

You can see packages making use of SummarizedExperiment (Depend/Import) on the [SummarizedExperiment homepage](https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html), some shown in the screenshot below.


```{r SE-packages,  echo=FALSE, out.width = '80%'}
knitr::include_graphics("../fig/summarizedexperiment_packages.png")
```

### Creating a SummarizedExperiment

We will demonstrate what SummarizedExperiment is by creating one. We will use the `rna` dataset again.

We start by loading the tidyverse packages which we will use to import and format the data.

```{r load-package, message=FALSE}
library("tidyverse")
```

```{r next-steps_data, echo = FALSE, message = FALSE}
rna <- read_csv("data/rnaseq.csv")
```

Load the `rna` data.

```{r load-data, eval=FALSE}
rna <- read_csv("course-data/data/GSE96870/rnaseq.csv")
```

From this table we will create the 3 different tables we need to create a SummarizedExperiment.
 
- **An expression matrix**

```{r}
# Select the columns needed and convert to wide format
counts <- rna %>%
  select(gene, sample, expression) %>%
  pivot_wider(names_from = sample,
              values_from = expression)

# Convert to matrix format
count_matrix <- counts %>% 
    select(-gene) %>% 
    as.matrix()

# Set gene ids as rownames
rownames(count_matrix) <- counts$gene

# Take a look
count_matrix[1:5, ]
dim(count_matrix)
```

- **A table describing the samples**

```{r}
# Select the columns needed 
sample_metadata <- rna %>%
  select(sample, organism, age, sex, infection, strain, time, tissue, mouse)

# Remove redundancy
sample_metadata <- unique(sample_metadata)

# Take a look
sample_metadata
```

- **A table describing the genes**

```{r}
# Select the columns needed 
gene_metadata <- rna %>%
    select(gene, ENTREZID, product, ensembl_gene_id, external_synonym, chromosome_name, gene_biotype, phenotype_description, hsapiens_homolog_associated_gene_name)

# Remove redundancy
gene_metadata <- unique(gene_metadata)

# Take a look
gene_metadata
```

We will create a `SummarizedExperiment` from these tables:

- The count matrix that will be used as the **`assay`**

- The table describing the samples will be used as the **sample metadata** slot

- The table describing the genes will be used as the **features metadata** slot

To do this we can put the different parts together using the
`SummarizedExperiment` constructor:

```{r, message = FALSE}
#BiocManager::install("SummarizedExperiment")
library("SummarizedExperiment")
```

To create the `SummarizedExperiment` object we need to provide inputs for the arguments `assays` (count matrix), `colData` (sample metadata) and `rowData` (gene metadata). Because `assays` can accept multiple assays (e.g. raw counts, log normalized) we put them in a list (SimpleList) and give each assay a name (we'll use counts). We can see we need to do this in the help for the SummarizedExperiment function `?SummarizedExperiment `.

```{r}
se <- SummarizedExperiment(assays = SimpleList(counts=count_matrix),
                           colData = sample_metadata,
                           rowData = gene_metadata)
se
```

We will save this SummarizedExperiment object in case we need it later.

```{r}
save(se, file = "data/SE.rda")
```

### Exploring a SummarizedExperiment

If we check size of our data in tabular format

```{r}
print(object.size(rna), units = "MB")
```

And compare to our data in SummarizedExperiment format

```{r}
print(object.size(se), units = "MB")
```

We can see SummarizedExperiment format has reduced the size ~ 5 fold, providing a very efficient way to store the data.


Using this data structure, we can access the expression matrix with
the `assay` function:

```{r}
head(assay(se))
dim(assay(se))
```

We can access the sample metadata using the `colData` function:

```{r}
colData(se)
dim(colData(se))
```

We can also access the feature metadata using the `rowData` function:

```{r}
head(rowData(se))
dim(rowData(se))
```


### Subsetting a SummarizedExperiment

SummarizedExperiment can be subset just like with data frames,
with numerics, characters or logicals. 

Importantly, the sample and feature are coordinated. If we need to exclude one sample from the assay, it will be automatically removed from the sample metadata in the same operation. 

Let's try that. Imagine we've discovered we need to exclude GSM2545336 as it's a failed sample. We can do that with below.

```{r}
se_good <- se[, se$sample != "GSM2545336"]
```

We can check the dimensions and see that we now have 21 samples instead of 22.

```{r}
dim(se_good)
```

If we take a look at the sample metadata we can see GSM2545336 is no longer there.

```{r}
colData(se_good)
```

If we take a look at the counts we can see GSM2545336 is no longer there.

```{r}
head(assay(se_good))
```

To make a smaller dataset to examine further, we will subset our `se` object. We will create a new instance of class SummarizedExperiment that contains only the first 5 features (genes) for the first 3 samples.

```{r}
se1 <- se[1:5, 1:3]
se1
```

```{r}
colData(se1)
rowData(se1)
```

We can also use the `colData()` function to subset on something from the sample metadata, or the `rowData()` to subset on something from the feature metadata.
For example, here we keep only miRNAs and the non infected samples:

```{r}
se1 <- se[rowData(se)$gene_biotype == "miRNA",
          colData(se)$infection == "NonInfected"]
se1
assay(se1)
colData(se1)
rowData(se1)
```


> ## Challenge
>
> Extract the gene expression levels of the first 3 genes in samples at time 0 and at time 8.
>
> > ## Solution
> >
> >
> > ```{r, purl=FALSE}
> > assay(se)[1:3, colData(se)$time != 4]
> >
> > # Equivalent to
> > assay(se)[1:3, colData(se)$time == 0 | colData(se)$time == 8]
> > ```
> {: .solution}
{: .challenge}

#### Adding variables to metadata

We can also add information to the metadata.
Suppose that you want to add the center where the samples were collected...

```{r}
colData(se)$center <- rep("University of Illinois", nrow(colData(se)))
colData(se)
```

This illustrates that the metadata slots can grow indefinitely without affecting
the other structures!

## tidySummarizedExperiment

You may be wondering, can we use tidyverse commands to interact with SummarizedExperiment objects. The answer is yes, we can with the tidySummarizedExperiment package.

Remember what our SummarizedExperiment object looks like.

```{r, message = FALSE}
se
```

Load tidySummarizedExperiment and then take a look at the se object again.

```{r, message = FALSE}
#BiocManager::install("tidySummarizedExperiment")
library("tidySummarizedExperiment")

se
```

It's still a SummarizedExperiment object, so maintains the efficient structure, but now we can view it as a tibble. Note the first line of the output says this, it's a SummarizedExperiment-tibble abstraction. We can also see in the second line of the output the number of transcripts and samples. 

If we want to revert to the standard SummarizedExperiment view we can do that.
```{r}
options("restore_SummarizedExperiment_show" = TRUE)
se
```

But here we will use the tibble view.

```{r}
options("restore_SummarizedExperiment_show" = FALSE)
se
```

We can now use tidyverse commands to interact with the SummarizedExperiment object.

We can use `slice` to choose rows by number e.g. to choose the first row.

```{r}
se %>% 
    slice(1)
```

We can use `filter` to filter for rows using a condition e.g. to remove a failed sample.

```{r}
se %>% 
    filter(.sample != "GSM2545336")
```

We can use `select` to specify columns we want to view. 

```{r}
se %>% 
    select(.sample)
```

```{r}
se %>%
    count(.sample)
```

We can use `distinct` to see what distinct sample information we have.

```{r}
se %>%
    distinct(.sample, infection, sex)
```

We can use `mutate` to add metadata info.

```{r}
se %>%
    mutate(center = "University of Melbourne")
```

We can use `unite` to combine multiple columns into a single column.

```{r}
se %>%
    unite("group", c(infection, time))
```

We can also combine commands with the tidyverse pipe `%>%`.

For example, we could combine `group_by` and `summarise` to get the total counts for each sample.

```{r}
se %>%
    group_by(.sample) %>%
    summarise(total_counts=sum(counts))
```

We can treat `se` as a normal tibble for plotting.

Here we plot the distribution of counts per sample.

```{r tidySE-plot}
se %>%
    ggplot(aes(counts + 1, group=.sample, color=infection)) +
    geom_density() +
    scale_x_log10() +
    theme_bw()
```


> ## Challenge
> > 1. Extract the miRNA and NonInfected this time using tidyverse commands
> >
> > ## Solution
> >
> > ```{r, purl=FALSE}
> > se %>% 
> > filter(gene_biotype == "miRNA" & infection == "NonInfected")
> > ```
> {: .solution}
> >
> >
> > 2. Extract the gene expression levels of the 3 first genes in samples at time 0 and at time 8 this time using tidyverse commands
> >
> > ## Solution
> >
> > ```{r, purl=FALSE}
> > se %>% 
> >   filter(time == 0 | time == 8) %>% 
> >   group_by(.sample) %>% 
> >   slice(1:3)
> > ```
> {: .solution}
{: .challenge}


## Extensions to SummarizedExperiment

### SingleCellExperiment

The [SingleCellExperiment](https://bioconductor.org/packages/devel/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html) container extends SummarizedExperiment. It follows similar conventions but is for single-cell data. Rows represent features (genes, transcripts, genomic regions) and columns represent cells. It provides methods for storing dimensionality reduction results and data for alternative feature sets (e.g., synthetic spike-in transcripts, antibody-derived tags). It is the central data structure for Bioconductor single-cell packages.

```{r SCE, echo=FALSE, out.width = '80%'}
knitr::include_graphics("https://bioconductor.org/books/release/OSCA.intro/images/SingleCellExperiment.png")
```

The [Orchestrating Single-Cell Analysis with Bioconductor book](https://bioconductor.org/books/release/OSCA/) is a great resource for single cell analysis that describes using SingleCellExperiment.

As there is tidySummarizedExperiment for interacting with SummarizedExperiment objects using tidyverse commands, there is equivalently the [tidySingleCellExperiment](https://stemangiola.github.io/tidySingleCellExperiment/) package for SingleCellExperiment objects.

### Interactive exploration

[iSEE](https://bioconductor.org/packages/release/bioc/vignettes/iSEE/inst/doc/basic.html) is a Bioconductor package that provides an interactive graphical user interface for exploring data stored in SummarizedExperiment and SingleCellExperiment objects.

```{r iSEE, echo=FALSE, out.width = '80%'}
knitr::include_graphics("https://raw.githubusercontent.com/iSEE/iSEE/master/vignettes/screenshots/basic-demo.png")
```

